x-pack/plugins/reporting/server/browsers/chromium/driver_factory/test.ts

```diff
@@ -0,0 +1,159 @@
+  let browserLogs = '';
+  let closed = false;
+
+  const browserProcess = spawn(binaryPath, finalArgs, {
+    // On non-windows platforms, `detached: true` makes child process a
+    // leader of a new process group, making it possible to kill child
+    // process tree with `.kill(-pid)` command. @see
+    // https://nodejs.org/api/child_process.html#child_process_options_detached
+    detached: process.platform !== 'win32',
+  });
+
+  return new Promise((resolve) => {
+    const cleanup = async (err?: Error) => {
+      if (closed) return;
+      if (err) {
+        logger.error(`Browser didn't start or didn't bind to the local port in time`);
+        logger.error(err);
+      }
+      const help: BrowserDiagnosticPayload['help'] = [];
+      closed = true;
+      clearTimeout(timer);
+
+      // Sometimes logs with significant issues can creep in after browser has bound it's remote
+      // debugging port, so let them flow through before SIGKILL is sent.
+      await sleep(browserCoolDown);
+
+      browserProcess.removeAllListeners();
+      rl.removeAllListeners();
+      rl.close();
+
+      for (const knownIssue of BROWSER_TEST_COMMON_ISSUES.keys()) {
+        const helpText = BROWSER_TEST_COMMON_ISSUES.get(knownIssue);
+        if (browserLogs.includes(knownIssue) && helpText) {
+          logger.warn(helpText);
+          help.push(helpText);
+        }
+      }
+
+      if (browserProcess && browserProcess.pid && !browserProcess.killed) {
+        try {
+          browserProcess.kill('SIGKILL');
+        } catch (error) {
+          logger.warn(
+            `Reporting was unable to close the browser process (PID: ${browserProcess.pid}). Please check your open processes and ensure that the browser processes that Reporting has launched have been killed.\nError cause: ${error.stack}`
+          );
+        }
+      }
+
+      logger.debug(`deleting chromium user data directory at [${userDataDir}]`);
+
+      del(userDataDir, { force: true }).catch((error) => {
+        logger.error(`error deleting user data directory at [${userDataDir}]!`);
+        logger.error(error);
+      });
+
+      return resolve({
+        help,
+        logs: browserLogs,
+        success: !err && !help.length,
+      });
+    };
+
+    const rl = createInterface({ input: browserProcess.stderr });
+
+    const timer = setTimeout(
+      () => cleanup(new Error(`Browser didn't bind successfully in ${maxTimeToWait} milliseconds`)),
+      maxTimeToWait
+    );
+
+    rl.on('line', (data) => {
+      logger.debug(`Browser log output: "${data}"`);
+      browserLogs += data + '\n';
+      const match = data.match(/^DevTools listening on/);
+      if (match) {
+        logger.debug(`Browser successfully started remote DevTools listener`);
+        return cleanup();
+      }
+    });
+
+    browserProcess.once('exit', () =>
+      cleanup(new Error(`Browser exited abnormally during startup`))
+    );
+  });
```

Comment:
I got a little lost in reading this file, because I start reading the body of `cleanup` before really seeing what the function actually does, and `rl` and `timer` are referenced in it before I know what those variables are.

I also found it a little hard to follow that the `resolve` call is inside of `cleanup`.

What do you think about:
 - move `cleanup` below everything.
 - instead of `resolve` inside of cleanup, you can call resolve after the cleanup promise is done, like:
    ```js
     cleanup().then(resolve)
    ```